### ISSUE TITLE
[Guide] Expected Disk consumption

### ISSUE THREAD

**Issue by cocoon** [Devolutions/WaykDen-ps#65]
Hi!
Just started to try to follow the guide and my first real problem I faced was that my Win 2019 VM (40GB C:) run out of disk space while it was pulling the docker images as they are huge.

I think a short hint about the expected amount of required disk space would be good before starting to pull.

Maybe also how to redirect the default location, I added now another virtual disk and created manually the config file as it was not there by default, stopped docker service and started again and the new directory was used directly, folder structure was created automatically when starting docker service again:
https://github.com/docker/for-win/issues/185

Cheers!

**Comment by awakecoding** [comment]
Thanks for the tip, I had never looked into changing the docker paths for Windows! Indeed, Windows containers based on server core do take a lot of space (~5GB for the base image). We intend to eventually get nano server containers, but we'd have to do it for all containers to really get the benefit, including a third-party containers. Otherwise, you'd have a few nano server containers (~200MB) alongside server core containers (~5GB). In the meantime, just using more disk space works :)

**Comment by awakecoding** [comment]
Would 64GB be a good minimum baseline? or even more?

**Comment by cocoon** [comment]
I don't know about the growing, I think you have a better overwiew about what is a minimum and how it could grow over time? But at least any number might help to see what to expect at the beginning.

I am still fighting to get it running at all. I didn't know where I could ask, I didn't want to create new issues until I expect it really to be an issue, or if the Forum would be good for this.

Currently as soon as I add LDAP config, it fails to start. 
```
Starting den-server
C:\Program Files\Docker\docker.exe: invalid reference format: repository name must be lowercase.
See 'C:\Program Files\Docker\docker.exe run --help'.
Error: No such object: den-server
Error: No such object: den-server
Error: No such object: den-server
Error: No such object: den-server
Error starting den-server
At C:\Program Files\WindowsPowerShell\Modules\WaykDen\2020.1.4\Public\WaykDenService.ps1:466 char:9
+         throw "Error starting $($Service.ContainerName)"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Error starting den-server:String) [], RuntimeException
    + FullyQualifiedErrorId : Error starting den-server
```


I even compiled my own WaykDen-ps and extended logging to see the output of the commands, some kind of Logging would be helpful ;)


I don't know if the spaces in the docker vars are escaped correctly, for example if you follow the example with the AD Group:

```
-e LDAP_USER_GROUP=Domain Users -e JET_SERVER_URL=
```


But even with a group without spaces it just fails to start den-server and I currently don't know where to start digging to find the root cause.

**Comment by awakecoding** [comment]
I noticed the problem last Friday for the spaces in the environment variable value, I can make a quick fix by changing how the docker command is generated. We hadn't noticed it yet because we were using a "wayk-users" group, but most Windows group names have spaces, so it is indeed a problem. We are also updating our LDAP integration such that it can be configured from the web ui rather than from the cmdlet.

As for LDAP integration with Active Directory, please make sure that you have LDAPS configured, as it will become the recommended connection option with the upcoming LDAP group policy changes. Our current LDAP implementation does not support NTLM signing + encryption over LDAP, and it also does not support the channel binding token (CBT) over LDAPS when using the spnego bind type (NTLM/Kerberos). However, the simple bind type works just fine over LDAPS, and it becomes a viable option with the TLS secure transport as opposed to LDAP. LDAPS will therefore become the way to integrate with Active Directory. We made a small mistake of using the spnego bind type by default in the last release, so LDAPS works as long as you don't enforce the channel binding token. The next release will change the default bind type to the simple bind when using LDAPS.

I recommend reading this excellent blog article on how to set up LDAPS in Active Directory: https://astrix.co.uk/news/2020/1/31/how-to-set-up-secure-ldap-for-active-directory

Setting up LDAPS looked much harder than it really is

**Comment by awakecoding** [comment]
As for the proxy settings in Wayk Now, did you already have https_proxy and http_proxy environment variables, which is why you had to clear them? Our code is supposed to check the registry on Windows to detect proxies, but I'd have to check if we pick up the environment variables anyway if they are set. Can you give me more details on the proxy exceptions list and where it is normally stored in the Windows registry? Our current implementation detects the proxy servers, but it doesn't go far into using exclusion lists, unfortunately.

**Comment by cocoon** [comment]
Ah great good to know it would be a cause for problems!
I just configured LDAPS today again and tested the connection with ldp.exe, have the certificates imported in the den server.

Btw there is a patch announced from Microsoft that enforces LDAPS around summer.

I just found this sentence, maybe you could add this param to the exmaple, so one might not miss the sentence like me ;)

```
It is important to specify the server IP since there is not DNS resolution in the docker container. 
```

**Comment by cocoon** [comment]
Proxy:

Yes for Docker I had to be able to pull the images I had to define the global system environment variables.

So for the client I had to clear them manually right before launching the client from the same cmd.
The Wayk Client (expecially curl or whatever it was) is picking up the environment variables, in the past I always had to define them before launching the client, but this way I can successfully use Wayk behind the proxy.

The exclusion list is needed in my environment, because the proxy cannot access our internal IPs/DNS-Names so the exceptions are needed.

To make it just a bit more challenging for you, there is another type of proxy settings available that you set with "netsh winhttp" ^^

https://securelink.net/en-be/insights/windows-proxy-settings-explained/
https://candritzky.wordpress.com/2011/08/11/winhttp-proxy-settings/

The bypass list is configured in "ProxyOverride"
```
Computer\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Internet Settings\
ProxyOverride="10.x.*.*;10.Y.*.*;*.my.FQ.DN.com;<local>"
```

**Comment by cocoon** [comment]
**Suggestion:** 
A custom option area in the client where one can use a dropdown to select the source where to import from (all the various found proxy settings in the system) or to provide custom ones just for the Wayk Client.

Then the Wayk Client should set it's own env variables like http_proxy and no_proxy for its context.

Regarding the parsing, it should support:

- Wildcard Domains like: ```*.contoso.com```

- Networks like ```192.168.0.0/24```

- IE Wildcards: ```10.0.*.*;<local>```

- Maybe offer RegEx ones



On Linux one can for example simply exclude all IPs like this that I collected:
```
export no_proxy=0,1,2,3,4,5,6,7,8,9
export no_proxy=`echo 10.1.1.{1..255} | sed 's/ /,/g'`
export no_proxy="`echo 10.1.1.{1..254},` 10.1.1.255"
export no_proxy="`echo 10.1.{1..255}.{1..254},` 10.1.255.255"
```

**Comment by cocoon** [comment]
LDAP: 
I cannot get it to work.

I even installed the Active Directory role onto the same server, setup Certificates, LDAPS tested with ldp.exe, bind with the new user account is working in ldp.exe, but whatever I try to configure with Set-WaykConfig, it won't start den-server.

Tried many possible things like:
LdapServerType was always set as: ActiveDirectory

LdapUsername:
```
wayk-ro@domain.something
wayk-ro
 "CN=WaykRO,CN=Users,DC=remote,DC=lab"
```
LdapPassword: always the same without spaces or sepcial chars

LdapServerUrl:
```
ldap://Server-FQDN
ldap://Server-FQDN:636
ldaps://Server-FQDN:636

LDAPS://Server-FQDN:636
LDAP://Server-FQDN:636
LDAP://Server-FQDN:636
LDAP://Server-FQDN

ldap://Server-IP
ldaps://Server-IP:636
```

LdapServerIp: always the one and only public Ip 10.x.x.x, not the docker ones

Once tried with:
LdapBaseDn "CN=Users,DC=mydom,DC=lab"

**Comment by cocoon** [comment]
Aaaah OK progress ... I removed the proxy from docker now, as the error was about accessing "http://den-lucid" ... I hate Proxies sooo much!


So now pulling of course fails when starting, but as I already have all images locally this is no real problem.
I can get a better error message after adding LDAP config, I will try again:

```
PS C:\Den> docker logs den-server
2020-03-24T13:25:20.314266700+01:00 INFO den_server::router::router - Redis is not configured. Cache in memory will be used.
2020-03-24T13:25:20.316269+01:00 WARN den_server::router::router - Keen client not configured.
2020-03-24T13:25:20.316269+01:00 INFO den_server::router::router - Nats is not configured.
2020-03-24T13:25:20.316269+01:00 INFO den_server::router::router - Starting listeners on ws://0.0.0.0:4491
2020-03-24T13:25:20.319267400+01:00 DEBUG reqwest::connect - starting new connection: http://den-lucid:4242/
2020-03-24T13:25:20.698286900+01:00 DEBUG reqwest::async_impl::client - response '200 OK' for http://den-lucid:4242/client
2020-03-24T13:25:20.817268100+01:00 DEBUG reqwest::async_impl::client - response '200 OK' for http://den-lucid:4242/client/7W5Q0Frt2IMLPgbNjZp6GoCe
2020-03-24T13:25:21.892268500+01:00 INFO den_server::model - Default user has not been created. At least one user exist.
2020-03-24T13:25:21.895268300+01:00 INFO den_server::http::http_server - Loading http controllers
thread 'main' panicked at 'Den-Server cannot start without fully initializing its services: AccountServiceError(AdAccountError(Other("Ldap bind failed (sasl_spnego_bind): rc=49 (invalidCredentials), dn: \"\", text: \"80090308: LdapErr: DSID-0C090585, comment: AcceptSecurityContext error, data 57, v4563\u{0}\"")))', src\libcore\result.rs:1188:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

**Comment by awakecoding** [comment]
We still have to improve the error handling for LDAP integration, as the only way you'll see that it doesn't work is through "docker logs den-server" at this point. We intended to rework the LDAP integration and make it configurable through the Web UI rather than through cmdlet parameters, which would hopefully let us do a "test connection parameters" button.

If you have the GPO enforcing the LDAP channel binding token, the current release will not work. We've tested LDAPS with a fresh test domain controller and it worked, but it didn't work with our production domain controller that has the GPO. In both cases, I would recommend that you make it work with certificate validation disabled first, then enable certificate validation and make sure you have the correct LdapServerUrl + trusted root CA. The trusted root CA can be imported using the Import-WaykLdapCertificate command, which takes as input just the root CA (not including the intermediate) in PEM format. If DNS resolving does not work correctly (it can happen for many reasons), the workaround to get the correct name is to use the expected common name in the LdapServerUrl + provide the real domain controller IP address in LdapServerIp.

Here is my suggestion: rather than wait for the webui-based LDAP configuration, we could work on making a small incremental release this week that changes the default bind type over LDAPS from the spnego bind type to the simple bind type. We've tested it, and it works correctly with the new LDAP channel binding token GPO. It would also become the official way to support Active Directory as we don't expect the LDAP (not LDAPS) integration to work with the GPO enforcing NTLM signing + encryption over LDAP, another feature which we do not support.

If you are fine with this suggestion, we'll fast track a small release to this week so that you can at least get unblocked on the Active Directory integration.

**Comment by cocoon** [comment]
Yes that sound great, thanks for all the help.

**Comment by awakecoding** [comment]
@cocoon we've just made a small incremental release with fixes for the Active Directory LDAPS integration. We've made the bind type configurable (Simple, Spnego), but the default now over LDAPS is the simple bind type which works correctly and safely with the latest GPOs.

https://www.powershellgallery.com/packages/WaykDen/2020.1.5

We don't have a nice way of testing LDAP parameters other than launching Wayk Den and checking the den-server logs for now. We'll improve on that once it is configurable through the Web UI, but in the meantime you should be able to get it working through the cmdlet parameters. Let me know if you encounter any issues.

**Comment by cocoon** [comment]
I nailed down the cause for the following error:
```
error sending request for url (http://den-lucid:4242/client): error trying to connect: dns error: No such host is known.
```

It was NOT the Proxy, this is something funny I think.
As I said I installed the ActiveDirectory +DNS role directly onto the Docker Host.

What Windows is doing automatically is to enter the following DNS-Server into the Windows Servers IP-Config for the network interface:
127.0.0.1

I think Docker dosn't work well with this, I expect that docker or the containers are trying to use the same IP for resolving DNS, so what is happening is that they try to ask 127.0.0.1 and that is the container itself so they can't get a response.

You can correct me if that is wrong, but as soon as I swapped 127.0.0.1 with the real Servers IP and rebootet, it was working again.

Maybe this could end up as a hint in an FAQ ^^
... yes it is NOT best practice to use 127.0.0.1 as DNS-Server ...

I did all this now in a local VM from scratch at home where no proxy server is involded at all and wondered, why I got this error again ;)

**Comment by awakecoding** [comment]
In this case, 'den-lucid' is expected to resolve to the docker network IP address of the 'den-lucid' container (our OAuth login service). This is something that cannot be resolved on the host, only within containers of the same docker network. If you use Start-WaykDen -Verbose, we print out the full docker run commands with all of its parameters. I haven't tried specifying DNS options for the containers, but I know there are a couple of options available to control it.

However, I'm surprised that the internal docker network DNS resolving got messed up in this way. When you say you installed the Active Directory + DNS roles, you mean you installed Active Directory Domain Services + the DNS server on the host, hoping to use the host as your domain controller for testing, or something else entirely?

**Comment by awakecoding** [comment]
By the way, be careful about 127.0.01 in the case of containers, because 127.0.0.1 inside a container always means the localhost of the internal docker network, never the localhost of the docker host itself. Using the host IP address should work, however.

**Comment by cocoon** [comment]
> you installed Active Directory Domain Services + the DNS server on the host, hoping to use the host as your domain controller for testing

Yes I want to get LDAP working somehow.  Still trying.

I have not much experience with docker up to now, so I still try to learn how everything is working.

**Comment by cocoon** [comment]
And I got LDAP working here with JumpCloud and could login with a user in the web interface, but that won't work behind a proxy ^^

**Comment by awakecoding** [comment]
If you got it to work with JumpCloud, I don't see why the proxy would prevent it from working. The LDAP integration is done between Wayk Den and the LDAP server, Wayk Now remains completely unaware of what is used behind the scenes. Thank you for your perseverance, it is really helpful.

If you need a good reference for Docker on Windows, I'd recommend the excellent book "Docker on Windows" (https://www.packtpub.com/virtualization-and-cloud/docker-windows-second-edition) by Elton Stoneman, otherwise there are a bunch of really useful guides available for free on the author's blog: https://blog.sixeyed.com/

Regarding the DNS issue, is the DNS server on the host literally '127.0.0.1', and if so, did you manually configure it that way? There's a good chance that if you just configure the DNS server to use the host IP address rather than 127.0.0.1, it would work within the containers. Even if they are inside a Docker network, we've seen that it was possible to reach the docker host using its full IP address from the host network.

**Comment by cocoon** [comment]
**The Good News:**
LDAPS is working now with ActiveDirectory 👍

```
jet_relay_url: https://api.jet-relay.net
jet_server_url: 
ldap_server_url: ldaps://myserver.mydomain.xxx:636
ldap_server_ip: 10.0.x.x
ldap_username: MYADUSER
ldap_password: MYPASSWORD
ldap_user_group: Domain Users 
ldap_server_type: ActiveDirectory
ldap_bind_type: simple
```

**The best news:** 
I even got the client to connect successfully to it. 🎉


**The not so good news:**
The name resolving of the service-names or container-names still seems to often break and I don't know why.
I found many reported issues about this topic, but I already checked that the DNS-Client-Cache-Timout is set to 0, I even extended the WaykDen-ps-Classes with the following:

**File: WaykDen\Private\DockerHelper.ps1**
```
class DockerService
{
    [string] $Image
    [string] $Platform
    [string] $ContainerName
    [string] $HostName  <-- ADDED THIS

...

DockerService([DockerService] $other) {
        $this.Image = $other.Image
        $this.Platform = $other.Platform
        $this.ContainerName = $other.ContainerName
        $this.HostName = $other.HostName <-- ADDED THIS
```


**File: WaykDen\Public\WaykDenService.ps1**
```
# den-mongo service
    $DenMongo = [DockerService]::new()
    $DenMongo.ContainerName = 'den-mongo'
    $DenMongo.HostName = 'den-mongo' <-- ADDED THIS

...
# den-picky service
    $DenPicky = [DockerService]::new()
    $DenPicky.ContainerName = 'den-picky'
    $DenPicky.HostName = 'den-picky' <-- ADDED THIS

# den-lucid service
    $DenLucid = [DockerService]::new()
    $DenLucid.ContainerName = 'den-lucid'
    $DenLucid.HostName = 'den-lucid' <-- ADDED THIS

# den-server service
    $DenServer = [DockerService]::new()
    $DenServer.ContainerName = 'den-server'
    $DenServer.HostName = 'den-server' <-- ADDED THIS

if ($ServerCount -gt 1) {
        1 .. $ServerCount | % {
            $ServerIndex = $_
            $Instance = [DockerService]::new([DockerService]$DenServer)
            $Instance.ContainerName = "den-server-$ServerIndex"
	    $Instance.HostName = "den-server-$ServerIndex" <-- ADDED THIS

# den-traefik service
    $DenTraefik = [DockerService]::new()
    $DenTraefik.ContainerName = 'den-traefik'
    $DenTraefik.HostName = 'den-traefik' <-- ADDED THIS

...

function Get-DockerRunCommand
{
    [OutputType('string[]')]
    param(
        [DockerService] $Service
    )

    $cmd = @('docker', 'run')

    $cmd += @('--name', $Service.ContainerName)
	
	if($Service.HostName) <-- ADDED THIS
	{
		$cmd += @('--hostname', $Service.HostName)
	}
```


**That helped somehow to do something like this:**
```
docker exec den-lucid ping den-mongo
docker exec den-lucid ping -4 den-mongo
```

**But in between caused a new problem with connecting to the DB:**
```
some error from den-sevrer regarding without a DB it cannot start ...
```

But right now it just works, and I leave it running for now  ^^


I get errors on the Dashboard and I cannot register the client, both throw errors:

Dashboard after login with no computer registered:
```
Operation forbidden
Server error - /wayk/problems/forbidden

The operation is forbidden
```

WaykNow-Client when clicking on the Button "Register this computer for unattended access"
```
Title: Wayk Now Authentication
You do not have the permission to perform that action
```

**Comment by awakecoding** [comment]
Thanks a lot for the tips regarding the DNS issue! We have another issue who reported similar issues, so I'll look into adding the --hostname parameter and try it out. I will also do further research on the docker hostname / DNS resolution parameters and see what might have an impact.

Regarding the permission problem for  "Register this computer for unattended access", you need to be logged in with a user that has the "Machine Contributor" role assignment:

![den_user_action_menu](https://user-images.githubusercontent.com/295841/77650440-e2fcd800-6f41-11ea-97a5-1823a9ae4f19.png)
![den_user_role_dford_machine_contributor](https://user-images.githubusercontent.com/295841/77650457-e7c18c00-6f41-11ea-8f87-8ed50d02d61b.png)

This procedure will be improved in the future through the usage of a pre-generated enrollment token that will authorize the registration of a machine with Wayk Den. This enrollment token will also facilitate automated enrollment with a script.

**Comment by cocoon** [comment]
Thanks for the hints ... I just feel that I missed it somewhere ... where or how can I assign roles to users ... I can login with my AD-User but is there something like a default admin that can login that I missed? 

Just came around another small problem of the Wayk Now Client:
If I for whatever reason would have to use another port for the HTTPS-Service of Wayk-Den, lets say 5443 and I confiugre the client to use Den-Connection to WSS://den.x.y:5443 it wokrs for the discovery, but in the debug logs I see that it still tries to get the configuration using explicit port 443 not 5443. Maybe something to think about.

For now I just could go for the normal port 443.

**Comment by cocoon** [comment]
And Regarding your question here:

> If you got it to work with JumpCloud, I don't see why the proxy would prevent it from working. The LDAP integration is done between Wayk Den and the LDAP server, Wayk Now remains completely unaware of what is used behind the scenes. 

The problem is not the client, it is the server that is behind a proxy and I did not find a way to use Den --> LDAPS --> HTTP-Proxy --> JumpCLoud ;)

**Comment by cocoon** [comment]
**"White Labeling"**
is working, too :)

Suggestion: move from 7zip to either additional branding.zip support or only .zip, as the Dashboard-Export is coming as .zip not as .7z, might be easier for customers

**Comment by awakecoding** [comment]
@cocoon we are finalizing the Wayk Now 2020.1.4 release that switches to .zip for the branding file, which should make things much simpler in the future. Hopefully it'll come out today

**Comment by cocoon** [comment]
Here is the error I got sometimes when using the "--hostname" option for all containers:

**docker logs den-server**
```
2020-03-26T16:57:12.892588400+01:00 ERROR r2d2 - No servers available for the provided ReadPreference.
thread 'main' panicked at 'Cannot start Den-Server without a database: "couldn\'t create r2d2 connection pool: timed out waiting for connection: No servers available for the provided ReadPreference."', src\libcore\result.rs:1188:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

I then just started den-server again:
```
docker start den-server
```

And it worked, maybe the DB just needed more time or something like that ...

**Comment by cocoon** [comment]
Got my first Argo-Tunnel running with cloudflared, I can connect somehow, but not authenticate:

```
2020-03-27T07:36:01.004395500+01:00 DEBUG cowrpc::transport::r#async::websocket - << Http Request
2020-03-27T07:36:01.005885300+01:00 DEBUG reqwest::async_impl::client - response '200 OK' for http://den-lucid:4242/openid/.well-known/openid-configuration
2020-03-27T07:36:01.007350600+01:00 DEBUG reqwest::async_impl::client - response '200 OK' for http://den-lucid:4242/auth/jwks
2020-03-27T07:36:01.007350600+01:00 ERROR den_server::http::utils - Cannot authenticate user: couldn't fetch public key from lucid: fetched jwk set is empty
2020-03-27T07:36:01.007350600+01:00 INFO den_server::http::middlewares::log - POST /wayk/register 400 Bad Request
2020-03-27T07:36:01.007350600+01:00 DEBUG cowrpc::transport::r#async::websocket - >> Http Response
```

**Comment by cocoon** [comment]
OK with HTTP instead of HTTPS it is working 👍 

```
-ListenerUrl http://den.mydomain.xyz
```

instead of 

```
-ListenerUrl https://den.mydomain.xyz
```

**Comment by cocoon** [comment]
Next success: it is so much easier and smaller with Ubuntu 18.04.4 ... no problems at all, was using my Windows-Den-AD-Server for LDAP without a problem ^^

Next challenge: get all this running from behind the proxy ... I expect I need the proxy for "JetRelayUrl https://api.jet-relay.net"

**Comment by awakecoding** [comment]
@cocoon the ListenerUrl / ExternalUrl configuration appears to be a major pain point for everybody, I'm thinking I'll have to add a "Test-WaykDen" command to run a bunch of validation checks.

As for Ubuntu vs Windows Server, you are correct, the Linux containers are MUCH smaller. We'll eventually port our Windows containers to nano server instead of server core, which should bring down the size a lot. Windows containers are the minority in the docker world, but they are the best way to get real Windows-native support, otherwise we'd really have to tell people to use Linux containers on Linux.

Regarding the proxy, if you can manage to avoid using a proxy in front of the relay servers, it would definitely help for performance. You can also try setting up your own relay servers instead of using ours, it gives you a fully isolated Wayk Den deployment with no reliance on our services.

**Comment by awakecoding** [comment]
@cocoon I'd like to reproduce the intermittent DNS issues you've observed on Windows, because I get a feeling it appears to be related to Windows containers specifically. Basically, all you did was create a Windows Server 2019 virtual with the latest updates, that's all? Is the host configured with external DNS servers or not?

**Comment by cocoon** [comment]
@awakecoding Hi,
regarding: 

> You can also try setting up your own relay servers instead of using ours, it gives you a fully isolated Wayk Den deployment with no reliance on our services

What needs to talk to the relay server? I would expect it must be reachable from the clients and server?

And for the other problem the configuration should be something like this currently and still fails to start, so I could still help to debug:

Windows 2019 Server
Running Active Directory + DNS Server
In the Windows own IP settings it has as only entry the public IP address of the server
The DNS Server is Forwarding to another DNS Server for records it cannot resolv
All Proxy Settings removed so no access to the internet

Current state (+ broken ldp.exe to LDAPS for some reason I don't know):
```
PS C:\den> docker logs den-server
2020-03-27T11:27:08.165683400+01:00 INFO den_server::router::router - Redis is not configured. Cache in memory will be used.
2020-03-27T11:27:08.166676600+01:00 WARN den_server::router::router - Keen client not configured.
2020-03-27T11:27:08.166676600+01:00 INFO den_server::router::router - Nats is not configured.
2020-03-27T11:27:08.166676600+01:00 INFO den_server::router::router - Starting listeners on ws://0.0.0.0:4491
2020-03-27T11:27:08.169676500+01:00 DEBUG reqwest::connect - starting new connection: http://den-lucid:4242/
2020-03-27T11:27:09.427643800+01:00 ERROR den_server - Unable to initialize DenServer: "error sending request for url (http://den-lucid:4242/client): error trying to connect: dns error: No such host is known. (os error 11001)"
PS C:\den> docker ps -a
CONTAINER ID        IMAGE                                               COMMAND                    CREATED             STATUS                           PORTS               NAMES
575916f30eba        devolutions/den-server:1.15.0-servercore-ltsc2019   "c:\\wayk\\den-server.…"   6 hours ago         Exited (0) 6 hours ago                               den-server
900d043b7107        devolutions/den-lucid:3.6.5-servercore-ltsc2019     "c:\\wayk\\lucid.exe"      6 hours ago         Up 6 hours (unhealthy)           4242/tcp            den-lucid
93d0cb3d4140        devolutions/picky:4.2.1-servercore-ltsc2019         "c:\\wayk\\picky-serve…"   6 hours ago         Up 6 hours                       1234/tcp            den-picky
5163c0696dff        mongo:4.2-windowsservercore-1809                    "mongod --bind_ip_all"     6 hours ago         Up 6 hours                       27017/tcp           den-mongo
e780e5767672        a4919c09dd96                                        "/traefik --file --c…"     3 days ago          Exited (3221225786) 3 days ago                       den-traefik
```

**Comment by awakecoding** [comment]
@cocoon the relay servers need to be reachable by Wayk Now (both the client and server) and Wayk Den. It is used for peer-to-peer connectivity between a Wayk Now client and Wayk Now server, and works using two outbound connections to a given relay server. It attempts opening a TCP or WSS transport, and selects the faster one that works.

**Comment by cocoon** [comment]
Thanks for the details, I thought that it works like this just from the name, but was not sure.

But another good news is, I got it working behind my proxy with Ubuntu!

But not with Argo / cloudflared, that has no support for HTTP Proxies:
https://github.com/cloudflare/cloudflared/issues/110

Instead I used https://ngrok.com/ successfully.

**Comment by cocoon** [comment]
If you want to troubleshoot directly on my broken Windows 2019, I could give you Wayk Now Access next week?

**Comment by awakecoding** [comment]
we can take a look next week if you'd like, but I think I'd better try and reproduce the issue locally :)

**Comment by cocoon** [comment]
sure if you like and succeed to reproduce it in your environment, less work for me :)

Interesting: ngrok tunnel works from behind the proxy ... but the proxy blocks it if I access it with a browser or Wayk Now 🙈

**Comment by si458** [comment]
I’ll just tag onto this if I may

Have you thought about doing a step by step of setting the den up on Ubuntu?

Only reason is I spun up a Ubuntu container with 20gb space and installed docker, .net core sdk and PowerShell and run the den on that and have had no issues at all so far!

Also saves a Windows licence/premium from cloud providers :)

**Comment by cocoon** [comment]
Hi, if I would not be forced to use a http-proxy, it would be all so easy.

With Ubuntu I had the problem, that cloudflared has problems with systemd-resolv, so I disabled it and used resolv.conf:
https://askubuntu.com/questions/907246/how-to-disable-systemd-resolved-in-ubuntu

And that cloudflared doesn't work with proxy (only DNS not dialing edge).

My options that I am thinking about are:

- host an alternative to argo in the cloud somewhere
- use Azure with one Ubuntu-VM + JumpCloud or similar

**Comment by si458** [comment]
sounds madness setup that,

My setup was simple
1. ran a Ubuntu vm
2. installed docker, powershell, .net core sdk, certbot
3. created certificates using certbot as certonly
4. built latest powershell scrips from github repo, 
5. imported the scripts and setup the den using the certificates from certbot
6. Setup a single container using newest Traefik 2 and the certs from certbot to proxy everything to localhost:4000 and away it was

so simple and easy! Again, no issues so far

**Comment by awakecoding** [comment]
@cocoon I agree about the Ubuntu-base step-by-step guide, I'll look into it. I started with a Windows-based guide because this is what most of our users are familiar with, but I can tell at this point it appears to be 50-50 between Windows and Linux in terms of preference. Linux is by far simpler if people are used to it, as opposed to Windows (better Docker support, smaller containers, etc).

**Comment by awakecoding** [comment]
@si458 awesome! I'm glad you could get it working :) Is your Ubuntu VM in the cloud, or on-premises? I'm curious about certbot, I haven't tried it. Alternatively, there is built-in support for letsencrypt in traefik that I didn't get a chance to try it yet. I did document a quick procedure to obtain a letsencrypt certificate using Posh-ACME: https://github.com/Devolutions/WaykDen-ps/blob/master/docs/getting-started-acme.adoc

**Comment by si458** [comment]
@awakecoding the ubuntu vm runs on my kit in a friends datacenter (the cloud in a sense) but not on-premises
i had to use certbot because my dns in again run by me on my plesk node, not on aws, azure, etc
just managed to get the jet-relay working too on the ubuntu vm

**Comment by cocoon** [comment]
@awakecoding One thing I would like to try would be to use the ListenerUrl insinde the network for the clients that would need to talk over a Proxy.

I expect this is not working if I use http://xyz I need to use https://.

I think I got this working once with windows.
But for some reason I don't get it running in Ubuntu, I even can not access the listeneurl with the firefox browser running on the same Ubuntu.

I use a selfsigned certificate. Could you give some stepts, mabye even openssl commands how to be able to use https-listener and access this URL instead of the external one for all clients that are inside the same network?

And I was only ever able to import pfx-format certificates, got always base64-errors when trying to import pem-files + key.

Thanks!


**My Ubuntu Setup**

**I configured listenerurl:**
```
Set-WaykDenConfig -ListenerUrl https://wayk18.remote.lab
Import-WaykDenCertificate -Password 0000 -CertificateFile /root/wayk18.remote.lab_certificate.pfx
```


**I imported the ca-certificates like this:**
```
apt-get install ca-certificates
cp *.crt /usr/share/ca-certificates
dpkg-reconfigure ca-certificates
```

**And I still get this error in Firefox:**
```
err_ssl_unrecognized_name_alert
```

**And this error in cloudflared:**
error="Get https://wayk18.remote.lab: remote error: tls: alert(112)"

**OpenSSL:**
```
root@wayk18:~# openssl s_client -showcerts -connect wayk18.remote.lab:443 </dev/null
CONNECTED(00000005)
139680494952896:error:14094458:SSL routines:ssl3_read_bytes:tlsv1 unrecognized name:../ssl/record/rec_layer_s3.c:1528:SSL alert number 112
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 319 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

**Comment by cocoon** [comment]
Update to the broken Windows instance with the name resolution problem:
I have now the workaround that I:

- added the --hosatname to the docker containers 

- and installed mongoDB on Windows and use it with my servers fqdn:
```
Set-WaykDenConfig -MongoExternal $true -MongoUrl 'mongodb://den.mydom.com:27017'
```

**Comment by awakecoding** [comment]
@cocoon self-signed certificates probably won't work, at a bare minimum you would need a custom CA with a root and an intermediate CA. We don't provide a way of disabling certificate validation in Wayk Now, instead we enforce using the system-provided certificate stores. I would advise against rolling your own CA, there are so many pitfalls on that path and you'll need to make your custom CA trusted on all machines where Wayk Now is used.

**Comment by cocoon** [comment]
Oh I can tell it works, and sure, maybe I would need to import certificates etc, but hey, now with MongoDB running on the Windows-AD-Den-Server it looks good.

The only problem is that the WaykNow Client cannot use the ListenerUrl it must be the ExternalUrl.
Any chance to make it possible that the WaykNow Client can connect to the ListernerUrl if that is different to the external URL?

**Comment by awakecoding** [comment]
@cocoon no, one should **always** use the external url at all times. Our authentication microservice is very sensible to the external URL, and we expose various endpoint URLs for service detection that are based on the external url. Can you tell me why you'd like to use the listener url specifically?

**Comment by si458** [comment]
@awakecoding @cocoon i was able to use the listenerurl when connecting to my own waykden its was just less secure as no https, but i then ran my own traefik container in front of port 443 and proxied everything backwards to the den-traefik container on port 440
before - http://remote.mywebserver.com:4000
after - https://remote.mywebserver.com

**Comment by cocoon** [comment]
The reason for it is that the WaykNow-Unnatended Service would run in the same network and would be able to directly connect to the WaykDen.

Without it it must go the way out using the http_proxy and the external URL is xyz.ngrok.io. that is blocked by the proxy for example, so I cannot use it currently.

**Comment by awakecoding** [comment]
@cocoon so you'd like to use more than one "external urls", one for the local network, and one truly external on the internet, as a workaround for issues you have with having a proxy in front of ngrok.io? I have never tried using an http proxy in front of ngrok.io, I wonder if it would work with cloudflare argo tunnels.

**Comment by awakecoding** [comment]
@si458 putting your own traefik with https + port 443 in front of our traefik container should work just fine, as long as you're configuring Wayk Den with the correct final external url. As for accessing Wayk Den through the listener url, it is not supported. You may have partial success with it, but we are aware of specific problems that might arise with our authentication microservice in this case.

**Comment by cocoon** [comment]
@awakecoding OK, I will try different ways then.

**Comment by awakecoding** [comment]
@cocoon can this issue be closed? feel free to open as many separate issues as you wish

**Comment by cocoon** [comment]
@awakecoding yes this can be closed. Thanks 👍
